be/src/storage/primary_index.cpp

```diff
@@ -1173,6 +1185,28 @@ const Slice* PrimaryIndex::_build_persistent_keys(const Column& pks, uint32_t id
     }
 }
 
+const Slice* PrimaryIndex::_build_persistent_keys(std::vector<const Column*>& key_cols,
+                                                  std::vector<std::pair<uint32_t, uint32_t>>& idx_ranges,
+                                                  std::vector<Slice>* key_slices) const {
+    DCHECK(key_cols.size() == idx_ranges.size());
+    for (auto i = 0; i < key_cols.size(); i++) {
+        if (key_cols[i]->is_binary()) {
+            const Slice* vkeys = reinterpret_cast<const Slice*>(key_cols[i]->raw_data());
+            for (size_t j = idx_ranges[i].first; j < idx_ranges[i].second; j++) {
+                key_slices->emplace_back(vkeys[j]);
+            }
+        } else {
+            uint32_t idx_begin = idx_ranges[i].first;
+            const uint8_t* keys = key_cols[i]->raw_data() + idx_begin * _key_size;
+            for (size_t j = idx_ranges[i].first; j < idx_ranges[i].second; j++) {
+                key_slices->emplace_back(keys, _key_size);
+                keys += _key_size;
+            }
+        }
+    }
```

Comment:
if there are 10 rows, 3 key columns, there will be 30 key_slices? how can it be used in persistent_index insert/update methods?

Reply:
> if there are 10 rows, 3 key columns, there will be 30 key_slices? how can it be used in persistent_index insert/update methods?

No, autually the `key_cols` means segments. For each segment, we will combine all key columns and encoded them as a new column and one `key_col` means one segment.